internal/provider/waypoint/resource_waypoint_application.go

```diff
@@ -272,9 +441,77 @@ func (r *ApplicationResource) Read(ctx context.Context, req resource.ReadRequest
 		data.ReadmeMarkdown = types.StringNull()
 	}
 
+	inputVars, err := clients.GetInputVariables(ctx, client, data.Name.ValueString(), loc)
+	if err != nil {
+		resp.Diagnostics.AddError(err.Error(), "Failed to fetch application's input variables.")
+		return
+	}
+
+	resp.Diagnostics.Append(readInputs(ctx, inputVars, data, varTypes)...)
```

Comment:
there is indirection here that I think is unnecessary. Is there a reason to embed the `readInputs` function call inside this `Append`? What about `readInputs` makes it particularly suited to return diagnostics as opposed to simply formatting and returning two `[]*InputVar` and we do the `SetValueFrom` inline with this READ method, and do `resp.Diagnostics.AddError` there?

Reply:
Returning diagnostics (which is returned by the call to `types.SetValueFrom` within `readInputs`) helps reduce code duplication, because if we returned two `[]*InputVar` we'd need to make the `SetValueFrom` calls wherever `readInputs` is called. It's not a large amount of duplicated code though, so I'm OK with moving the `SetValueFrom` to the calling function side to reduce the unnecessary indirection.